core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,155 @@
+
+    clock.setTime(2000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1"))
+    // Executor 1 failed again at Time 00::00:02
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
+
+    clock.setTime(3000)
+    // Executor 2 failed at Time 00:00:03
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1", "2"))
+
+    clock.setTime(6000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1", "2"))
+
+    clock.setTime(8000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("2"))
+
+    clock.setTime(10000)
+    tracker.expireExecutorsInBlackList()
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
+  }
+
+  test("default strategy works") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled","false")
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf, clock)
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("1","2"))
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set())
+
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("2"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(taskInfo_3_3_hostB, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set("2"))
+    
+    tracker.updateFailureExecutors(taskInfo_3_3_hostB, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 3L) === Set("3"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(taskInfo_2_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
+  }
+
+  test ("threshold strategy works") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled","false")
+      .set("spark.scheduler.blacklist.strategy", "threshold")
+      .set("spark.scheduler.blacklist.threshold.maxFailureTaskNumber", "1")
+      .set("spark.scheduler.blacklist.threshold.maxBlackExecutorNumber", "0")
+
+    sc = new SparkContext(conf)
+    val scheduler = spy(new TaskSchedulerImpl(sc, 1))
+    when(scheduler.getExecutorsAliveOnHost("hostA")).thenReturn(Some(Set("1","2")))
+
+    val tracker = new BlacklistTracker(conf, clock)
+    tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 1L) === Set())
+    tracker.updateFailureExecutors(taskInfo_1_2_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, 2L) === Set("1", "2"))
+    assert(tracker.nodeBlacklist() === Set("hostA"))
```

Comment:
Sorry this is a pain, but I'd like you to make a few changes to increase test coverage a bit:

(a) add executor "3" to hostA
(b) change "maxBlackExecutorNumber" to 1
(c) after failing two tasks on executor 1, check that all tasks are blacklisted on executor 1, (regardless of task id):
```scala
tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
assert(tracker.executorBlacklist(scheduler, 1L) === Set())
tracker.updateFailureExecutors(taskInfo_1_1_hostA, FAILURE)
// executor 1 is now blacklisted from *all* tasks
(0 until 10).foreach { assert(tracker.executorBlacklist(scheduler, _) === Set("1")) }
assert(tracker.nodeBlacklist() === Set())
```

(d) then fail add more task failures, so that you blacklist executor 2, which will trigger the node blacklist so also blacklist executor 3.

(e) after removing executor 1 from the blacklist, then for all tasks, the blacklist is just executor 2 (executor 3 is also removed).

the point is to check that the separation of executor & node blacklists.